L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoSeeding_SA.cc

```diff
@@ -0,0 +1,232 @@
+    if (row + 1 <= int(config_.cRows() - 1)) {
+      *hc += (lHistogram[binIndex + config_.cColumns()] / 2);
+    }
+  }
+}
+
+// Simple threshold maxima finder
+// Keep only bins passing threshold
+// Threshold can vary vs r/z
+void HGCalHistoSeeding::thresholdMaximaFinder(HGCalHistogramCellSAPtrCollection& histogram) const {
+  const unsigned int stepLatency = config_.getStepLatency(Maxima2D);
+  for (auto& hc : histogram) {
+    hc->addLatency(stepLatency);
+    if (hc->S() <= config_.thresholdMaxima(hc->sortKey())) {
+      hc->setS(0);
+      hc->setX(0);
+      hc->setY(0);
+      hc->setN(0);
+    }
+  }
+}
+
+// Temporary simulation of local maxima finder
+// Not an emulation of any firmware
+void HGCalHistoSeeding::localMaximaFinder(HGCalHistogramCellSAPtrCollection& histogram) const {
+  // const std::vector<unsigned> maximaWidths{ 6, 5, 5, 5, 4, 4, 4, 4, 3, 3, 3, 3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 }; // Padded this with 4*1 at the end.  Firmware this was based on has 40 entries, but there are 44 bins.
+  const std::vector<unsigned> maximaWidths(config_.cRows(), 1);
+
+  for (auto& hc : histogram) {
+    if (hc->S() > 0) {
+      const int colRef = hc->index();
+      const int rowRef = hc->sortKey();
+      bool isMaxima = true;
+      const int phiRange = maximaWidths.at(hc->sortKey());
+      for (int colOffset = -1 * phiRange; colOffset <= phiRange; ++colOffset) {
+        const int col = colRef + colOffset;
+        if (col < 0 || col >= (int)config_.cColumns())
+          continue;
+        for (int rowOffset = -1; rowOffset <= 1; ++rowOffset) {
+          const int row = rowRef + rowOffset;
+          if (row < 0 || row >= (int)config_.cRows())
+            continue;
+          const unsigned int binIndex = config_.cColumns() * row + col;
+          const auto& bin = histogram.at(binIndex);
+
+          if (colOffset == 0 && rowOffset == 0)
+            continue;
+          else if ((col < colRef) || (col == colRef && rowOffset == -1) || (col == colRef + 1 && rowOffset == -1)) {
+            if (!(hc->S() >= bin->S()))
+              isMaxima = false;
+          } else {
+            if (!(hc->S() > bin->S()))
+              isMaxima = false;
+          }
+        }
+      }
+      if (!isMaxima) {
+        hc->setS(0);
+        hc->setX(0);
+        hc->setY(0);
+        hc->setN(0);
+      }
+    }
+  }
+}
+
+// Calculate average of phi and r/z (now labelled X and Y) of TCs within a bin
+// As performed in firmware, where the sum of TC phi or r/z is multiplied by 2^17/N_TCs (factors stored in a LUT)
+// Then result is shifted by 17 bits
+void HGCalHistoSeeding::calculateAveragePosition(HGCalHistogramCellSAPtrCollection& histogram) const {
+  const unsigned int stepLatency = config_.getStepLatency(CalcAverage);
+  for (auto& hc : histogram) {
+    hc->addLatency(stepLatency);
+    if (hc->N() > 0) {
+      unsigned int inv_N = int(round(1.0 * 0x1FFFF / hc->N()));
```

Comment:
- Magic number `0x1FFFF` to be turned into named constant
- Multiplication by `1.0` to be replaced by an explicit conversion to float?